src/state/routing/utils.ts

```diff
@@ -97,6 +112,19 @@ const parsePair = ({ reserve0, reserve1 }: V2PoolInRoute): Pair =>
     CurrencyAmount.fromRawAmount(parseToken(reserve1.token), reserve1.quotient)
   )
 
-function isV3Route(route: V3PoolInRoute[] | V2PoolInRoute[]): route is V3PoolInRoute[] {
-  return route[0].type === 'v3-pool'
+const genericPoolPairParser = (pool: V3PoolInRoute | V2PoolInRoute): Pool | Pair => {
+  return pool.type === 'v3-pool' ? parsePool(pool) : parsePair(pool)
+}
+
+function isV3Route(route: (V3PoolInRoute | V2PoolInRoute)[]): route is V3PoolInRoute[] {
+  return route.every((pool) => pool.type === 'v3-pool')
+}
+
+function isV2Route(route: (V3PoolInRoute | V2PoolInRoute)[]): route is V2PoolInRoute[] {
+  return route.every((pool) => pool.type === 'v2-pool')
+}
+
+function isMixedRoute(route: (V3PoolInRoute | V2PoolInRoute)[]): route is (V3PoolInRoute | V2PoolInRoute)[] {
+  /// Must have at least one V3 pool and one V2 pool
```

Comment:
This isn't comprehensive. Should you also check that every route is either V3 or V2?

Reply:
Should be resolved by new logic in `getRouteProtocol`